src/Qir/Execution/Tools/Executable/QirExecutableGenerator.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Quantum.Qir.Utility;
+
+namespace Microsoft.Quantum.Qir.Tools.Executable
+{
+    internal class QirExecutableGenerator : IQirExecutableGenerator
+    {
+        private readonly IClangClient clangClient;
+        private readonly ILogger logger;
+
+        public QirExecutableGenerator(IClangClient clangClient, ILogger logger)
+        {
+            this.clangClient = clangClient;
+            this.logger = logger;
+        }
+
+        public async Task GenerateExecutableAsync(FileInfo executableFile, DirectoryInfo sourceDirectory, IList<DirectoryInfo> libraryDirectories, IList<DirectoryInfo> includeDirectories, IList<string> linkLibraries)
+        {
+            // Wrap in a Task.Run because FileInfo methods are not asynchronous.
+            await Task.Run(async () =>
+            {
+                var binDirectory = executableFile.Directory;
+                logger?.LogInfo($"Creating binary directory at {binDirectory.FullName}.");
+                executableFile.Directory.Create();
+
+                // Copy all library contents to bin.
+                logger?.LogInfo("Copying library directories into the executable's folder.");
+
+                if (!sourceDirectory.Exists)
+                {
+                    logger?.LogWarning($"Cannot find source directory: {sourceDirectory.FullName}");
+                }
+
+                var libDirs = new List<string>();
+                foreach (var dir in libraryDirectories)
+                {
+                    if (!dir.Exists)
+                    {
+                        logger?.LogWarning($"Cannot find given directory: {dir.FullName}");
+                    }
+                    else
+                    {
+                        CopyDirectoryContents(dir, binDirectory);
+                        libDirs.Add(dir.FullName);
+                    }
+                }
+
+                var includeDirs = new List<string>();
+                foreach (var dir in includeDirectories)
+                {
+                    if (!dir.Exists)
+                    {
+                        logger?.LogWarning($"Could not find given directory: {dir.FullName}");
+                    }
+                    else
+                    {
+                        includeDirs.Add(dir.FullName);
+                    }
+                }
+
+                var inputFiles = sourceDirectory.GetFiles().Select(fileInfo => fileInfo.FullName).ToArray();
```

Comment:
Consider Directory.GetFiles(sourceDirectory.FullName). Shorter and without linq and lambdas.

Reply:
In this case, `Select` is used to transform the `FileInfo` array returned by `GetFiles()` into a `string` array. It doesn't really do anything that requires a search pattern.